### PR TITLE
Ignore unmatched `:ignore` attributes

### LINKED ISSUE #48: Make `:ignore` also ingore existance of an attribute altogether
# New Feature Proposal

## Description

Right now, we can ignore a class attribute when comparing two HTML strings. However, if we set up a test with multiple cases, where one test does not have a class at all, it does not get ignored and instead we get an error:
```csharp
[Theory]
[InlineData("<div class:ignore></div>", "<div class=\"hello\"></div>")]
[InlineData("<div class:ignore></div>", "<div></div>")] // Will fail
[InlineData("<div required:ignore></div>", "<div required></div>")]
[InlineData("<div required:ignore></div>", "<div required=\"\"></div>")]
[InlineData("<div required:ignore></div>", "<div></div>")] // Will fail
public void ShouldHaveNoErrors(string controlHtml, string testHtml)
{
    var diffs = DiffBuilder.Compare(controlHtml).WithTest(testHtml).Build();
    Assert.Empty(diffs);
}
```

## Background

When I'm testing blazor code I also test for `null` values. However, when the Razor engine sees `attr=null` it removes the attribute altogether. Although there are thousands of workarounds to come up with, I think it could be a cute little feature.
We have talked about this a bit in https://github.com/bUnit-dev/bUnit/issues/1751

Then it seemed that because there are also boolean attributes, it might be a nifty feature in how it is set up right now. However, now when I look at it again, I feel that the `:ignore` suffix should also ignore the attribute not being there altogether. The way attributes work in HTML is that they are false by default, adding them makes them true. If we want to ignore a boolean attribute, we don't care whether it is true or false. The way it's working right now, the `:ignore` suffix forces the boolean value to be true. I.e. the `:ignore` more works as an `:exists`.

## Specification

I have taken a quick look at the HTML standard and there are two big categories and some sub categories in attributes:
- Key-Value attributes
- Boolean attributes
Then there are some more niche attributes:
- Enumerated attributes
- CORS attributes
- Event handler attributes
- Aria attributes

But I think, from an XML standpoint, these are all key-value attributes

Therefore, I propose that `:ignore` also ignores an attribute not being there altogether. As an alternative (to prevent breaking changes) we can add an extra `:ignoreExistance`.

### PR DESCRIPTION
**Pull request by RiRiSharp**
# Types of Changes

## Prerequisites

Please make sure you can check the following two boxes:

- [x] I have read the **CONTRIBUTING** document
- [x] My code follows the code style of this project

## Contribution Type

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [x] Bug fix (non-breaking change which fixes an issue, please reference the issue id)
- [ ] New feature (non-breaking change which adds functionality, make sure to open an associated issue first)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed

## Description

Closes #48

### CODE REVIEW

File: src/AngleSharp.Diffing/Core/HtmlDifferenceEngine.cs [hunk 1]
```diff
@@ -160,7 +161,13 @@ void MarkSelectedSourcesAsMatched(in AttributeComparison comparison)
...
         {
-            Context.MissingAttributeSources.AddRange(controls.GetUnmatched());
+            // Filter out unmatched :ignore attributes, they were meant to be ignored after all
+            // https://github.com/AngleSharp/AngleSharp.Diffing/issues/48
+            var controlsUnmatched = controls
+                .GetUnmatched()
+                .Where(c => !IgnoreAttributeComparer.IsIgnoreAttribute(c.Attribute));
+
+            Context.MissingAttributeSources.AddRange(controlsUnmatched);
             Context.UnexpectedAttributeSources.AddRange(tests.GetUnmatched());
```
[the remark this answers is not in the document]
Comment (RiRiSharp) [reply src/AngleSharp.Diffing/Core/HtmlDifferenceEngine.cs:171]: I have now created an extra test, tomorrow I will try to make it succeed

[the remark this answers is not in the document]
Comment (RiRiSharp) [reply src/AngleSharp.Diffing/Core/HtmlDifferenceEngine.cs:171]: That's actually quite funny, I'm not sure if I was awake or asleep but that's how I solved it in my head last night too. Making a fix now

[the remark this answers is not in the document]
Comment (RiRiSharp) [reply src/AngleSharp.Diffing/Core/HtmlDifferenceEngine.cs]: Fix is pushed now, basically what you wrote was correct. Awaiting your approval now

File: src/AngleSharp.Diffing/Strategies/AttributeStrategies/IgnoreAttributeComparer.cs
[the remark this answers is not in the document]
Comment (egil) [reply src/AngleSharp.Diffing/Strategies/AttributeStrategies/IgnoreAttributeComparer.cs:1]: Or maybe merge the two files into one static file, e.g. `IgnoreAttributeStrategy.cs`, have the `Compare` and `Match` methods in the static class, and then have the static `IsIgnoreAttribute` method there being shared.

File: src/AngleSharp.Diffing/Core/SourceMap.cs [hunk 2]
```diff
@@ -70,7 +70,7 @@ public IEnumerable<AttributeComparisonSource> GetUnmatched()
     {
         foreach (var source in _sources.Values)
         {
-            if (!_matched.Contains(source.Attribute.Name))
+            if (IsUnmatched(source.Attribute.Name))
```
[the remark this answers is not in the document]
Comment (egil) [reply src/AngleSharp.Diffing/Core/SourceMap.cs:73]: this is wrong. The method is right above the GetUnmatched() method

File: src/AngleSharp.Diffing/Strategies/AttributeStrategies/IgnoreAttributeComparer.cs
[the remark this answers is not in the document]
Comment (RiRiSharp) [reply src/AngleSharp.Diffing/Strategies/AttributeStrategies/IgnoreAttributeComparer.cs:1]: Do we need to keep `IgnoreAttributeComparer` for backwards compatibility?

**Comment by RiRiSharp** [comment]
I have:
- Merged `IgnoreAttributeComparer` and  `IgnoreAttributeMatcher` into `IgnoreAttributeStrategy`.
- Since `IgnoreAttrrubteComparer` is public, I have made it obsolete and made sure all code inside the project uses the new strategy class.
- Modified the pipeline for Linux a slight bit so hopefully the build does not fail anymore: https://github.com/NuGet/setup-nuget/issues/168
